### PR TITLE
feat(clickhouse): Add comprehensive support for INDEX, PROJECTION, OPTIMIZE TABLE, and AggregateFunction

### PR DESCRIPTION
**Pull request by leiyangyou**
# ClickHouse Enhancement: INDEX, PROJECTION, OPTIMIZE TABLE, and AggregateFunction Support

## Summary
This PR significantly enhances ClickHouse SQL support in SQLFluff by adding comprehensive support for advanced ClickHouse constructs that are commonly used in production environments.

### 🔍 INDEX Definitions
- **All ClickHouse index types**: `BLOOM_FILTER`, `MINMAX`, `SET`, `NGRAMBF_V1`, `TOKENBF_V1`, `HYPOTHESIS`
- **Parameterized indexes**: Support for `BLOOM_FILTER(0.01)` and similar constructs
- **Semantic parsing**: Uses `StringParser` with `type="index_type_identifier"` for proper categorization

### 📊 PROJECTION Definitions  
- **Table projections**: Full support for ClickHouse materialized projections
- **GROUP BY support**: Fixed critical parsing issue with GROUP BY clauses in projections
- **ORDER BY support**: Complete support for ORDER BY clauses in projections
- **No FROM clause**: Correctly handles projections without FROM clauses as per ClickHouse spec

### ⚡ OPTIMIZE TABLE Statements
- **Basic optimization**: `OPTIMIZE TABLE table_name`
- **Partition optimization**: `OPTIMIZE TABLE table_name PARTITION (expr1, expr2)`
- **Final optimization**: `OPTIMIZE TABLE table_name FINAL`
- **Deduplication**: `OPTIMIZE TABLE table_name DEDUPLICATE`

### 🧮 Enhanced AggregateFunction Data Type
- **Simple functions**: `AggregateFunction(count, UInt64)`
- **Parameterized functions**: `AggregateFunction(quantile(0.95), Float64)`
- **Multiple arguments**: `AggregateFunction(sum, Decimal(10, 2))`

### Key Fix: GROUP BY in Projections
The critical fix was adding `Sequence("GROUP", "BY")` to `SelectClauseTerminatorGrammar`. Previously, the SELECT clause parser was consuming the `GROUP` keyword as an alias for functions, causing `BY category` to become unparsable.

**Before**: `count(id) GROUP` (GROUP treated as alias) → `BY category` (unparsable)  
**After**: `count(id)` (proper termination) → `GROUP BY category` (properly parsed)

### New Segment Classes
- **IndexDefinitionSegment**: Comprehensive index parsing with all ClickHouse index types
- **ProjectionDefinitionSegment**: Table projection parsing with SELECT/GROUP BY/ORDER BY
- **OptimizeTableStatementSegment**: All OPTIMIZE TABLE variations
- **Enhanced DatatypeSegment**: Improved AggregateFunction support with FunctionSegment

### Grammar Enhancements
- **SelectClauseTerminatorGrammar**: Added GROUP BY terminator for projections
- **Index Type Parsing**: StringParser-based approach for semantic correctness
- **Function Parameter Support**: Enhanced AggregateFunction to handle parameterized functions

## Test Coverage
- **New test fixtures**: 4 comprehensive SQL test files with YML validation
- **12 new tests**: All passing with full coverage of new functionality
- **Regression testing**: All existing ClickHouse tests continue to pass
- **Real-world examples**: Tests based on actual ClickHouse documentation

### Test Files Added
```
test/fixtures/dialects/clickhouse/create_table_with_indexes.sql     - All index types
test/fixtures/dialects/clickhouse/create_table_with_projections.sql - Projection with GROUP BY  
test/fixtures/dialects/clickhouse/optimize_table.sql                - All OPTIMIZE variations
test/fixtures/dialects/clickhouse/aggregate_function_datatype.sql   - AggregateFunction examples
```

### INDEX Definitions
```sql
CREATE TABLE example (
    id UInt64,
    name String,
    category String,
    
    INDEX bloom_idx name TYPE BLOOM_FILTER(0.01) GRANULARITY 1,
    INDEX minmax_idx category TYPE MINMAX GRANULARITY 1,
    INDEX set_idx category TYPE SET(100) GRANULARITY 1
) ENGINE = MergeTree() ORDER BY id;
```

### PROJECTION Definitions
```sql
CREATE TABLE sales (
    date Date,
    product String,
    amount Decimal(10,2),
    
    PROJECTION product_summary (
        SELECT product, sum(amount), count() GROUP BY product
    )
) ENGINE = MergeTree() ORDER BY date;
```

### OPTIMIZE TABLE Operations
```sql
OPTIMIZE TABLE my_table;
OPTIMIZE TABLE my_table PARTITION (2023, 1);  
OPTIMIZE TABLE my_table FINAL;
OPTIMIZE TABLE my_table DEDUPLICATE;
```

### AggregateFunction Data Types
```sql
CREATE TABLE aggregates (
    id UInt64,
    user_count AggregateFunction(count, UInt64),
    quantile_data AggregateFunction(quantile(0.95), Float64)
) ENGINE = MergeTree() ORDER BY id;
```

## Impact
This enhancement brings SQLFluff's ClickHouse support much closer to production-ready status, enabling proper parsing of:
- Advanced table definitions with indexes and projections
- Data engineering workflows with table optimization  
- Complex analytical data types used in ClickHouse applications

## Validation
- ✅ All new functionality tested and passing
- ✅ No regressions in existing ClickHouse tests
- ✅ Grammar follows ClickHouse official documentation  
- ✅ Comprehensive edge case coverage

🤖 Generated with [Claude Code](https://claude.ai/code)

Co-Authored-By: Claude <noreply@anthropic.com>

### CODE REVIEW

File: src/sqlfluff/dialects/dialect_clickhouse.py [hunk 1]
```diff
@@ -1059,6 +1074,85 @@ class ColumnConstraintSegment(BaseSegment):
...
+                    Ref("ExpressionSegment"),  # index expression
+                    delimiter=Ref("CommaSegment"),
+                )
+            ),
+            # Without parentheses: INDEX name column1
+            Ref("ExpressionSegment"),
+        ),
+        "TYPE",
+        OneOf(
+            StringParser("MINMAX", CodeSegment, type="index_type_identifier"),
```
Comment (WittierDinosaur) [inline src/sqlfluff/dialects/dialect_clickhouse.py:1102]: You don't need to invoke StringParser here, you can just use "MINMAX" to reference the minmax keyword

**Comment by leiyangyou** [comment]
Didn't want to add minmax/etc as a keyword, the capitalisation rule on
keywords turns it into uppercase when configured.

How should it be done normally? As kerwords?

**Comment by akselbor** [comment]
Any news on when/if this will be merged in?

**Comment by WittierDinosaur** [comment]
@leiyangyou can you fix the pre-commit error, then lgtm

**Comment by peterbud** [comment]
Closing due to inactivity
